README
This is a small mastermind solver for a school project made by.

Xander van Doorn: misterx-89@hotmail.com
Rick Poot: rlpoot@gmail.com
Micha van Doorne: mail@doomic.nl


Changelog
commit 2d70279e9e65ae6168f7ddeaa79e99f57f9e635c
Author: Rick Poot <rpoot@chopper.localdomain>
Date:   Thu Oct 28 08:38:13 2010 +0200

    Presentatie gemaakt

commit afe90fc0d38b16d901930a9dbf40cbe84fff2fc6
Author: xander <mister-x89@hotmail.com>
Date:   Wed Oct 27 21:12:12 2010 +0000

    lastversion

commit 49f929ce8eacff609592963088f18c34fcfdbe68
Author: xander <mister-x89@hotmail.com>
Date:   Wed Oct 27 15:48:06 2010 +0000

    jippie

commit 37b531b4a57fab75dd1224a0f602e8ca00dc7205
Merge: 9913a88 1a12eb9
Author: xander <mister-x89@hotmail.com>
Date:   Wed Oct 27 12:11:08 2010 +0000

    Merge branch 'master' of github.com:rickpt/mastermind

commit 9913a88cd9361bef890a7ed86b812a8162cda9c1
Author: xander <mister-x89@hotmail.com>
Date:   Wed Oct 27 12:10:43 2010 +0000

    ?

commit 135d2d751b102266af371df8c514cffd682dd9a6
Merge: 11524ce 5515d01
Author: xander <mister-x89@hotmail.com>
Date:   Wed Oct 27 11:35:44 2010 +0000

    Merge branch 'master' of github.com:rickpt/mastermind

commit 11524ce84ec47a42cdf750da347a24cbfd959580
Merge: 6d74676 4f0e415
Author: xander <mister-x89@hotmail.com>
Date:   Wed Oct 27 11:30:52 2010 +0000

    improve history

commit 6d74676c5b1523b5fd36013a26ee73e7aab25aa6
Author: xander <mister-x89@hotmail.com>
Date:   Wed Oct 27 11:28:15 2010 +0000

    improve history

commit 1a12eb975374897a7ae78fa882ed75ac109713f8
Author: Rick Poot <rpoot@chopper.localdomain>
Date:   Wed Oct 27 11:55:48 2010 +0200

    	modified:   makefile

commit 7241d81950a0ca3881460e05e47bc8b7e471d856
Author: Rick Poot <rpoot@chopper.localdomain>
Date:   Wed Oct 27 11:54:54 2010 +0200

    	modified:   README

commit 5515d019dd62fd62fa9527db17159b41a17ae1b1
Author: Rick Poot <rpoot@chopper.localdomain>
Date:   Wed Oct 27 11:32:23 2010 +0200

    Updated .gitignore

commit 4f0e415f5bf4440401a18d01821c9d8cda69c002
Author: Rick Poot <rpoot@chopper.localdomain>
Date:   Wed Oct 27 11:24:48 2010 +0200

    added printfiletable() and pintfilecode

commit c46c546404e7d19d30c0f78c0a7d56b28974e7c4
Author: Rick Poot <rpoot@chopper.localdomain>
Date:   Wed Oct 27 10:57:20 2010 +0200

    remover vimrw

commit 61b297ab349e359839e71990919417aa0a5e5f84
Author: Rick Poot <rpoot@chopper.localdomain>
Date:   Wed Oct 27 10:51:39 2010 +0200

    Revert "Revert "vimrc""
    
    This reverts commit 53906af7989a34a232e7bec3bb720492ca0ad712.

commit ae1d6c984b1b2a05a0dfe7af784edf0e3c36610f
Author: Rick Poot <rpoot@chopper.localdomain>
Date:   Wed Oct 27 10:50:57 2010 +0200

    Revert "bump"
    
    This reverts commit 38fdb7c286252524fa5ec213b11bc3d1613c10a3.

commit 53906af7989a34a232e7bec3bb720492ca0ad712
Author: Rick Poot <rpoot@chopper.localdomain>
Date:   Wed Oct 27 10:45:35 2010 +0200

    Revert "vimrc"
    
    This reverts commit 563d3aa54a675d48e6c9e70dbc9f7979d897f6ac.

commit 38fdb7c286252524fa5ec213b11bc3d1613c10a3
Author: Rick Poot <rpoot@chopper.localdomain>

common.h
#include <stdio.h>
#include <fstream>


#ifndef __COMMON
#define __COMMON

//--------------------------------------------------------------------------------------------------------------definitions

enum eColor { empty, a=1, b=2, c=3, d=4, e=5, f=6 };
enum ePin   { none, white, black      };

typedef eColor Code [4];
typedef ePin   Pins [4];

struct History { Code code; Pins pins; };
History m_history[50];
int historypointer = -1;

void printCode(Code);
void printPins(Pins);
void printLine();
void printTab();
void printHistory();

void addRecordToHistory(Code, Pins);

void generateTable(); 

Code m_table [1300];
int tablepointer = 0; 

std::ofstream myfile;
void printFileCode(Code cd);

//--------------------------------------------------------------------------------------------------------------functions

void generateTable(){
	int counter= 0;
	for(int i=0; i<6; i++){
		for(int k=0; k<6; k++){
			for(int l=0; l<6; l++){
				for(int m=0; m<6; m++){
					m_table[counter][0] =  (eColor) (i+1);
					m_table[counter][1] =  (eColor) (k+1);
					m_table[counter][2] =  (eColor) (l+1);
					m_table[counter++][3] = (eColor) (m+1);
				}
			}
		}
	}
	tablepointer = counter;
}

void printTable(){
	for(int i=0; i< tablepointer; i++){
		printCode(m_table[i]);
		printLine();
	}
}

void printFileTable(){
    myfile.open("mastermind.log");
	for(int i=0; i< tablepointer; i++){
		printFileCode(m_table[i]);
	    myfile << "\n";	
	}
    myfile.close();
}

int countBlackPins(History *hs, int index){
	int count = 0;
	for(int i=0; i<4; i++){
		if(hs[index].pins[i] == black){
			count++;
		}
	}
	return count;
}

int countWhitePins(History *hs, int index){
	int count = 0;
	for(int i=0; i<4; i++){
		if(hs[index].pins[i] == white){
			count++;
		}
	}
	return count;
}

void printCode(Code cd){
	for( int i=0; i<4; i++){
		switch(cd[i]){
			case empty:
				printf("-");
			break;
			case a:
				printf("a");
			break;
			case b:
				printf("b");
			break;
			case c:
				printf("c");
			break;
			case d:
				printf("d");
			break;
			case e:
				printf("e");
			break;
			case f:
				printf("f");
			break;			
		}
	}	
}

void printFileCode(Code cd){
	for( int i=0; i<4; i++){
		switch(cd[i]){
			case empty:
				myfile << "-";
			break;
			case a:
				myfile << "a";
			break;
			case b:
				myfile << "b";
			break;
			case c:
				myfile << "c";
			break;
			case d:
				myfile << "d";
			break;
			case e:
				myfile << "e";
			break;
			case f:
				myfile << "f";
			break;			
		}
	}	
}

void printPins(Pins pn){
	for( int i=0; i<4; i++ ){
		switch(pn[i]){
			case none:
				printf("-");
			break;
			case white:
				printf("w");
			break;
			case black:
				printf("b");
			break;
		}
	}
}

void printLine(){
	printf("\n");
}

void printTab(){
	printf("\t");
}

void printHistory(){
	for(int i=0; i <= historypointer ; i++){
		printCode(m_history[i].code);
		printTab();
		printPins(m_history[i].pins);
		printLine();
	}
}

void addRecordToHistory(Code cd, Pins pn){
	int index = ++historypointer;
	m_history[index].code[0] = cd[0];
	m_history[index].code[1] = cd[1];
	m_history[index].code[2] = cd[2];
	m_history[index].code[3] = cd[3];
	m_history[index].pins[0] = pn[0];
	m_history[index].pins[1] = pn[1];
	m_history[index].pins[2] = pn[2];
	m_history[index].pins[3] = pn[3];	
}
#endif


validator.h
/**
 * important note:
 *
 * returns false if code is definetly not valid due to the heuristics. 
 * True means that the validator cannot detect if the code is valid or not. 
 * In other words, there can occur parodox situations which it's possible that a code is invalid but
 * this function returns a true either.
 *
 * */
bool validateCode(Code cd, History *hs){
	int whitePins= 0; int blackPins=0;
	int white, black;

/*	for(int i=0; i <= historypointer; i++){
		whitePins = countWhitePins(hs,i);
		blackPins = countBlackPins(hs,i);
		white = 0;
		black = 0;
	
		Code bckCode;
		bckCode[0] = hs[i].code[0];
		bckCode[1] = hs[i].code[1];
		bckCode[2] = hs[i].code[2];
		bckCode[3] = hs[i].code[3];
	
		Code bckCode2;
		bckCode2[0] = cd[0];
		bckCode2[1] = cd[1];
		bckCode2[2] = cd[2];
		bckCode2[3] = cd[3];
	
		for(int k=0; k<4; k++){
			if(bckCode[k] == bckCode2[k]){
				black++;
				bckCode2[k] = empty;	
				bckCode[k] = empty;
			}
		}
		for(int l=0; l<4; l++){
			if(bckCode[l] == bckCode2[0] && bckCode[l] != empty){
				white++;	
				bckCode2[0] = empty;
				bckCode[l] = empty;
			} else if(bckCode[l] == bckCode2[1] && bckCode[l] != empty){
				white++;
				bckCode2[1] = empty;
				bckCode[l] = empty;
			} else if(bckCode[l] == bckCode2[2] && bckCode[l] != empty){
				white++;
				bckCode2[2] = empty;
				bckCode[l] = empty;
			} else if(bckCode[l] == bckCode2[3] && bckCode[l] != empty){
				white++;
				bckCode2[3] = empty;
				bckCode[l] = empty;
			}	
		}
		if(white != whitePins || black != blackPins){
			return false;
		}
	}*/



	for(int i=0 ; i<=historypointer; i++){
		bool codeCan[4] = { true, true, true, true };
		bool  hisCan[4] = { true, true, true, true };
		whitePins = countWhitePins(hs,i);
		blackPins = countBlackPins(hs,i);
		white=0; black=0;
		for(int k=0; k<4; k++){
			if(cd[k] == hs[i].code[k]){
				black++;
				codeCan[k] = false;
				hisCan[k]  = false;
			}
		}
		for(int k=0; k<4; k++){
//			if(codeCan[k] == true){
				for(int l=0; l<4; l++){
					if(cd[k] == hs[i].code[l] && hisCan[l] == true && codeCan[k] == true){
						white++;
						hisCan[l] = false;
					}
				}
//			}

		} 
		if(white != whitePins || black != blackPins){
			return false;
		}
	}		
	return true;
}


main.cpp
#include <iostream>
#include <fstream>
#include <stdlib.h>
#include <GL/glut.h>
#include "imageloader.h"
#include "common.h"
#include "validator.h"


using namespace std;
float _Hangle = 165.0f;
float _Vangle = 100.0f;
float _x = 1.0f;
float _y = 0.0f;
float _z = -8.0f;
int _Rows = 8;
int _Zet = 0;
int _Position = 0;
float _ZMax = ((_Rows/2.0f)*1.2f)+0.2f;
float _positionColor = 0.0f;
bool _positionColorUp = true;
GLuint _textureBoard;
GLuint _textureManual;
bool _isAanzet = true;
bool _isLeftKeyPressed = false;
bool _isRightKeyPressed = false;
bool _isUpKeyPressed = false;
bool _isDownKeyPressed = false;
bool _isPlusKeyPressed = false;
bool _isMinKeyPressed = false;
bool _isAKeyPressed = false;
bool _isDKeyPressed = false;
bool _isSKeyPressed = false;
bool _isWKeyPressed = false;
bool _isEnterKeyPressed = false;
bool _isSpaceKeyPressed = false;
bool _isBackspaceKeyPressed = false;
int _balls[12][4];
int _pawns[12][4];

//ontvang alle antwoorden tot nu toe!
void addAnswer() {
	if(_Zet == 0){
		generateTable();
		_balls[_Zet][0] = (int) a; 
		_balls[_Zet][1] = (int) b;
		_balls[_Zet][2] = (int) c; 
		_balls[_Zet][3] = (int) d; 
	}
        else if(_Zet == 1){
		_balls[_Zet][0] = (int) b;
		_balls[_Zet][1] = (int) c;
		_balls[_Zet][2] = (int) d;
		_balls[_Zet][3] = (int) e;
		
	}
	else if(_Zet == 2){
		_balls[_Zet][0] = (int) c;
		_balls[_Zet][1] = (int) d;
		_balls[_Zet][2] = (int) e;
		_balls[_Zet][3] = (int) f;		
	} 
	if((_Zet-1) >=0) {
		Code cd; Pins pins;
		cd[0] = (eColor) _balls[_Zet-1][0];
		cd[1] = (eColor) _balls[_Zet-1][1];
		cd[2] = (eColor) _balls[_Zet-1][2];
		cd[3] = (eColor) _balls[_Zet-1][3];
		
		pins[0] = (ePin) _pawns[_Zet-1][0];
		pins[1] = (ePin) _pawns[_Zet-1][1];
		pins[2] = (ePin) _pawns[_Zet-1][2];
		pins[3] = (ePin) _pawns[_Zet-1][3];
		
		addRecordToHistory(cd,pins);
	}	
	bool isPossibility = true;
	if(_Zet > 2){
		isPossibility = false;
		for(int i=0; i<tablepointer; i++){
			if(m_table[i][0] != empty){
				if(!validateCode(m_table[i], m_history)){
					m_table[i][0] = empty;
				} else {					
					_balls[_Zet][0] = m_table[i][0];
					_balls[_Zet][1] = m_table[i][1];
					_balls[_Zet][2] = m_table[i][2];
					_balls[_Zet][3] = m_table[i][3];
					m_table[i][0] = empty;
					isPossibility = true;	
					break;
				}
			}
		}
	}
	_isAanzet = isPossibility;
} 

void handleKeypress(unsigned char key, int x, int y) {
	switch (key) {
		case 27: exit(0); break;
		case 43: _isPlusKeyPressed = true; break;
		case 45: _isMinKeyPressed = true; break;
		case 97: _isAKeyPressed = true; break;
		case 100: _isDKeyPressed = true; break;
		case 115: _isSKeyPressed = true; break;
		case 119: _isWKeyPressed = true; break;
	}
}

void handleKeyReleased(unsigned char key, int x, int y) {
	switch (key) {
		case 8: _isBackspaceKeyPressed = true; break;
		case 13: _isEnterKeyPressed = true; break;
		case 32: _isSpaceKeyPressed = true; break;
		case 43: _isPlusKeyPressed = false; break;
		case 45: _isMinKeyPressed = false; break;
		case 97: _isAKeyPressed = false; break;
		case 100: _isDKeyPressed = false; break;
		case 115: _isSKeyPressed = false; break;
		case 119: _isWKeyPressed = false; break;
	}
}

void handleSpecialKeypress(int key, int x, int y) {
	switch (key) {
		case GLUT_KEY_LEFT: _isLeftKeyPressed = true; break;
		case GLUT_KEY_RIGHT: _isRightKeyPressed = true; break;
		case GLUT_KEY_UP: _isUpKeyPressed = true; break;
		case GLUT_KEY_DOWN:_isDownKeyPressed = true; break;
	}
}

void handleSpecialKeyReleased(int key, int x, int y) {
	switch (key) {
		case GLUT_KEY_LEFT: _isLeftKeyPressed = false; break;
		case GLUT_KEY_RIGHT:_isRightKeyPressed = false; break;
		case GLUT_KEY_UP: _isUpKeyPressed = false; break;
		case GLUT_KEY_DOWN:_isDownKeyPressed = false; break;
	}
}

GLuint loadTexture(Image* image) {
	GLuint textureId;
	glGenTextures(1, &textureId);
	glBindTexture(GL_TEXTURE_2D, textureId);
	glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB, image->width, image->height, 0, GL_RGB, GL_UNSIGNED_BYTE,image->pixels);
	return textureId; //Returns the id of the texture
}


void initRendering() {
	glEnable(GL_DEPTH_TEST);
	glEnable(GL_COLOR_MATERIAL);
	glEnable(GL_LIGHTING);
	glEnable(GL_LIGHT0);
	glEnable(GL_LIGHT1);
	glEnable(GL_NORMALIZE);
	glClearColor(0.0f, 0.05f, 0.15f, 1.0f);

	Image* image = loadBMP("board.bmp");
	_textureBoard = loadTexture(image);
	delete image;
	Image* image2 = loadBMP("manual.bmp");
	_textureManual = loadTexture(image2);
	delete image2;
}

void handleResize(int w, int h) {
	glutFullScreen();
	glViewport(0, 0, w, h);
	glMatrixMode(GL_PROJECTION);
	
	glLoadIdentity();
	gluPerspective(45.0,(double)w / (double)h,0.1,300.0);
}

void drawTriangle(float topX, float topY,float topZ, float width, float height,float thickness) {
	glBegin(GL_TRIANGLES); 
	    //bottom
	    glNormal3f(0.0f, 1.0f, 0.0f);
	    glTexCoord2f(topX, topZ);
	    glVertex3f(topX, topY, topZ);
	    glTexCoord2f(topX-(width/2), topZ-height);
	    glVertex3f(topX-(width/2), topY, topZ-height);
	    glTexCoord2f(topX+(width/2), topZ-height);
	    glVertex3f(topX+(width/2), topY, topZ-height);

	    //top
	    glNormal3f(0.0f, 1.0f, 0.0f);
	    glTexCoord2f(topX, topZ);
	    glVertex3f(topX, topY+thickness, topZ);
	    glTexCoord2f(topX-(width/2), topZ-height);
	    glVertex3f(topX-(width/2), topY+thickness, topZ-height);
	    glTexCoord2f(topX+(width/2), topZ-height);
	    glVertex3f(topX+(width/2), topY+thickness, topZ-height);
	glEnd();
	glBegin(GL_QUADS); 
	   //front
	   glNormal3f(0.0f, 0.0f, -1.0f);
	   glTexCoord2f(topX-(width/2), topY);
	   glVertex3f(topX-(width/2), topY, topZ-height);
	   glTexCoord2f(topX-(width/2), +thickness);
	   glVertex3f(topX-(width/2), topY+thickness, topZ-height);
	   glTexCoord2f(topX+(width/2), topY+thickness);
	   glVertex3f(topX+(width/2), topY+thickness, topZ-height);
	   glTexCoord2f(topX+(width/2), topY);
	   glVertex3f(topX+(width/2), topY, topZ-height);

	   //left
	   glNormal3f(-0.5f, 0.0f, 0.5f);
	   glTexCoord2f(topX, topY);
	   glVertex3f(topX, topY, topZ);
	   glTexCoord2f(topX, topY+thickness);
	   glVertex3f(topX, topY+thickness, topZ);
	   glTexCoord2f(topX-(width/2), topY+thickness);
	   glVertex3f(topX-(width/2), topY+thickness, topZ-height);
	   glTexCoord2f(topX-(width/2), topY);
	   glVertex3f(topX-(width/2), topY, topZ-height);
	   
	   //right
	   glNormal3f(0.5f, 0.0f, 0.5f);
	   glTexCoord2f(topX, topY);
	   glVertex3f(topX, topY, topZ);
	   glTexCoord2f(topX, topY+thickness);
	   glVertex3f(topX, topY+thickness, topZ);
	   glTexCoord2f(topX+(width/2), topY+thickness);
	   glVertex3f(topX+(width/2), topY+thickness, topZ-height);
	   glTexCoord2f(topX+(width/2), topY);
	   glVertex3f(topX+(width/2), topY, topZ-height);
	glEnd();
}

void drawCube(float x1, float y1,float z1, float x2, float y2,float z2) {
	glBegin(GL_QUADS); 
	   //bottom
	   glNormal3f(0.0f, -1.0f, 0.0f);
	   glTexCoord2f(x1, z1);
	   glVertex3f(x1, y1, z1);
	   glTexCoord2f(x2, z1);
	   glVertex3f(x2, y1, z1);
	   glTexCoord2f(x2, z2);
	   glVertex3f(x2, y1, z2);
	   glTexCoord2f(x1, z2);
	   glVertex3f(x1, y1, z2);

	   //top
	   glNormal3f(0.0f, 1.0f, 0.0f);
	   glTexCoord2f(x1, z1);
	   glVertex3f(x1, y2, z1);
	   glTexCoord2f(x2, z1);
	   glVertex3f(x2, y2, z1);
	   glTexCoord2f(x2, z2);
	   glVertex3f(x2, y2, z2);
	   glTexCoord2f(x1, z2);
	   glVertex3f(x1, y2, z2);

	   //back
	   glNormal3f(0.0f, 0.0f, -1.0f);
	   glTexCoord2f(x1, y1);
	   glVertex3f(x1, y1, z2);
	   glTexCoord2f(x1, y2);
	   glVertex3f(x1, y2, z2);
	   glTexCoord2f(x2, y2);
	   glVertex3f(x2, y2, z2);
	   glTexCoord2f(x2, y1);
	   glVertex3f(x2, y1, z2);

	   //front
	   glNormal3f(0.0f, 0.0f, 1.0f);
	   glTexCoord2f(x1, y1);
	   glVertex3f(x1, y1, z1);
	   glTexCoord2f(x1, y2);
	   glVertex3f(x1, y2, z1);
	   glTexCoord2f(x2, y2);
	   glVertex3f(x2, y2, z1);
	   glTexCoord2f(x2, y1);
	   glVertex3f(x2, y1, z1);

	   //left
	   glNormal3f(-1.0f, 0.0f, 0.0f);
	   glTexCoord2f(z1, y1);
	   glVertex3f(x1, y1, z1);
	   glTexCoord2f(z1, y2);
	   glVertex3f(x1, y2, z1);
	   glTexCoord2f(z2, y2);
	   glVertex3f(x1, y2, z2);
	   glTexCoord2f(z2, y1);
	   glVertex3f(x1, y1, z2);

	   //right
	   glNormal3f(1.0f, 0.0f, 0.0f);
	   glTexCoord2f(z1, y1);
	   glVertex3f(x2, y1, z1);
	   glTexCoord2f(z1, y2);
	   glVertex3f(x2, y2, z1);
	   glTexCoord2f(z2, y2);
	   glVertex3f(x2, y2, z2);
	   glTexCoord2f(z2, y1);
	   glVertex3f(x2, y1, z2);
	  glEnd();
}

void drawBall(int position, int row, int color) {
	   glPushMatrix();
        glTranslatef( (2.5 - ((position - 1) * 1.2f)), 0.4, ((1.2f * (_Rows - 1))/2) - ((row - 1) * 1.2f));
		switch(color) {
		 case 1: glColor3f(1.0f, 1.0f, 1.0f); break;//white ball
		 case 2: glColor3f(0.1f, 0.1f, 0.1f); break;//black ball
		 case 3: glColor3f(0.7f, 0.0f, 0.0f); break;//red ball
		 case 4: glColor3f(0.0f, 0.7f, 0.0f); break;//green ball
		 case 5: glColor3f(0.0f, 0.0f, 0.7f); break;//blue ball
		 case 6: glColor3f(0.7f, 0.7f, 0.0f); break;//yellow ball
		}
	    glutSolidSphere(0.4, 25, 25);
	   glPopMatrix();
}

void drawPawn(int position, int row, int color) {
	   glPushMatrix();
	    float z = ((1.2f * (_Rows - 1))/2) - ((row - 1) * 1.2f);
		if(position == 1 || position == 2) {
			z = z + 0.3f;
		} else {
			z = z - 0.3f;
		}
		float x = -2.75f;
		if(position == 1 || position == 3) {
			x = -2.15f;
		}
        glTranslatef( x, 0.15, z );
		switch(color) {
		 case 1: glColor3f(1.0f, 1.0f, 1.0f); break;//white pawn
		 case 2: glColor3f(0.7f, 0.0f, 0.0f); break;//red pawn
		}
	    glutSolidSphere(0.15, 15, 15);
	   glPopMatrix();
}

void drawScene() {
	glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
	
	glMatrixMode(GL_MODELVIEW);
	glLoadIdentity();

	GLfloat ambientColor[] = {0.5f, 0.5f, 0.5f, 1.0f};
	glLightModelfv(GL_LIGHT_MODEL_AMBIENT, ambientColor);
	
	GLfloat lightColor0[] = {0.9f, 0.9f, 0.9f, 1.0f};
	GLfloat lightPos0[] = {-5.0f, 5.0f, -5.0f, 1.0f};
	glLightfv(GL_LIGHT0, GL_DIFFUSE, lightColor0);
	glLightfv(GL_LIGHT0, GL_POSITION, lightPos0);

	glPushMatrix();
	 glTranslatef(_x,_y,_z);
	 glRotatef(_Hangle,0.0f,1.0f,0.0f);
	 glRotatef(_Vangle,1.0f,0.0f,0.0f);
	 glEnable(GL_TEXTURE_2D);
	  glBindTexture(GL_TEXTURE_2D, _textureManual);
	  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
	  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
	  glColor3f(0.6f, 0.6f, 0.6f);
	  float x1 = -1.7f;
	  float y1 = -0.15f;
	  float z1 = 2.5f;
	  float x2 = 1.7f;
	  float y2 = -0.1f;
	  float z2 = -2.5f;
	  
	  glBegin(GL_QUADS); 
	   glNormal3f(0.0f, -1.0f, 0.0f);
	   glTexCoord2f(0.75f, 0.0f);
	   glVertex3f(x1, y1, z1);
	   glTexCoord2f(0.0f, 0.0f);
	   glVertex3f(x2, y1, z1);
	   glTexCoord2f(0.0f, 1.0f);
	   glVertex3f(x2, y1, z2);
	   glTexCoord2f(0.75f, 1.0f);
	   glVertex3f(x1, y1, z2);
	  glEnd();
	 glDisable(GL_TEXTURE_2D);

	 glEnable(GL_TEXTURE_2D);
	  glBindTexture(GL_TEXTURE_2D, _textureBoard);
	  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
	  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
	  glColor3f(0.6f, 0.6f, 0.6f);
	   //top
	   drawCube(-3.6f, -0.1f, _ZMax, 3.6f, 0.0f, _ZMax*-1.0f);
	   //front
	   drawCube(-3.1f, -0.6f, _ZMax-0.5f, 3.1f, -0.1f, _ZMax-0.6f);
	   //back
	   drawCube(-3.1f, -0.6f, (_ZMax-0.6f)*-1.0f, 3.1f, -0.1f, (_ZMax-0.5f)*-1.0f);

	   //draw topLayer
	  float _tZ = (_Rows/2.0f)*1.2f;
	  for(int i = 0; i < _Rows; i++) {
	   drawCube(-3.4f, 0.0f, _tZ, 3.4f, 0.3f, _tZ-0.15f);
	   drawCube(-3.4f, 0.0f, _tZ-1.05f, 3.4f, 0.3f, _tZ-1.2f);
	   drawCube(-3.4f, 0.0f, _tZ-0.15f, -2.9f, 0.3f, _tZ-1.05f);
	   drawCube(-2.6f, 0.0f, _tZ-0.15f, -2.3f, 0.3f, _tZ-1.05f);
	   drawCube(-2.9f, 0.0f, _tZ-0.45f, -2.6f, 0.3f, _tZ-0.75f);
	   drawCube(-2.3f, 0.0f, _tZ-0.45f, -2.0f, 0.3f, _tZ-0.75f);
	   glPushMatrix();//--------------------------------------------------------------
	    glRotatef(90.0f,0.0f,1.0f,0.0f);
	    drawTriangle((_tZ-0.6f)*-1, 0.0f, -1.7f, 0.9f, 0.3f, 0.3f); //end
	   glPopMatrix();
	   glPushMatrix();
	    glRotatef(180.0f,0.0f,1.0f,0.0f);
	    drawTriangle(1.7f, 0.0f, (_tZ-0.6f)*-1, 0.6f, 0.45f, 0.3f);
	   glPopMatrix();
	   drawTriangle(-1.7f, 0.0f, _tZ-0.6f, 0.6f, 0.45f, 0.3f);
	   glPushMatrix();
	    glRotatef(180.0f,0.0f,1.0f,0.0f);
	    drawTriangle(0.5f, 0.0f, (_tZ-0.6f)*-1, 0.6f, 0.45f, 0.3f);
	   glPopMatrix();
	   drawTriangle(-0.5f, 0.0f, _tZ-0.6f, 0.6f, 0.45f, 0.3f);
	   glPushMatrix();
	    glRotatef(180.0f,0.0f,1.0f,0.0f);
	    drawTriangle(-0.7f, 0.0f, (_tZ-0.6f)*-1, 0.6f, 0.45f, 0.3f);
	   glPopMatrix();
	   drawTriangle(0.7f, 0.0f, _tZ-0.6f, 0.6f, 0.45f, 0.3f);
	   glPushMatrix();
	    glRotatef(180.0f,0.0f,1.0f,0.0f);
	    drawTriangle(-1.9f, 0.0f, (_tZ-0.6f)*-1, 0.6f, 0.45f, 0.3f);
	   glPopMatrix();
	   drawTriangle(1.9f, 0.0f, _tZ-0.6f, 0.6f, 0.45f, 0.3f);
	   glPushMatrix();
	    glRotatef(180.0f,0.0f,1.0f,0.0f);
	    drawTriangle(-3.1f, 0.0f, (_tZ-0.6f)*-1, 0.6f, 0.45f, 0.3f);
	   glPopMatrix();
	   drawTriangle(3.1f, 0.0f, _tZ-0.6f, 0.6f, 0.45f, 0.3f);
	   glPushMatrix();
	    glRotatef(270.0f,0.0f,1.0f,0.0f);
	    drawTriangle(_tZ-0.6f, 0.0f, -3.1f, 0.9f, 0.3f, 0.3f); //end
	   glPopMatrix();
	   
	   _tZ = _tZ - 1.2f;
	  }
	  //huidige positie
	  if(_isAanzet) {
	   float z = ((1.2f * (_Rows - 1))/2) - ((_Zet) * 1.2f);
	   if(_Position == 0 || _Position == 1) {
		z = z + 0.3f;
	   } else {
		z = z - 0.3f;
	   }
	   float x = -2.75f;
	   if(_Position == 0 || _Position == 2) {
		x = -2.15f;
	   }
	   glColor3f(_positionColor, _positionColor, _positionColor);
	   drawCube(x + 0.15f, 0.0f, z + 0.15f, x - 0.15f, 0.01f, z - 0.15f);
	   if(_positionColorUp) _positionColor = _positionColor + 0.05f;
	   else _positionColor = _positionColor - 0.05f;
	   if(_positionColor > 1.0f) _positionColorUp = false;
	   if(_positionColor < 0.0f) _positionColorUp = true;
	  }

	 glDisable(GL_TEXTURE_2D);
	 int i = 0;
	 for(i=0; i<(_Zet + 1); i++) {
	  int j = 0;
	  for(j=0; j<4; j++) {
	   if(_balls[i][j] > 0 && _balls[i][j] <= 6) drawBall(j+1, i+1, _balls[i][j]);
	   if(_pawns[i][j] > 0 && _pawns[i][j] <= 2) drawPawn(j+1, i+1, _pawns[i][j]);
	  }
	 }
	glPopMatrix();
	glutSwapBuffers();
}

void update(int value) {
	//angle horizontal
	if(_isLeftKeyPressed) _Hangle += 2.0f;
	if(_isRightKeyPressed) _Hangle -= 2.0f;
	if (_Hangle > 360) _Hangle -= 360;
	if (_Hangle < 0) _Hangle += 360;
	//angle vertical
	if(_isUpKeyPressed) _Vangle += 2.0f;
	if(_isDownKeyPressed) _Vangle -= 2.0f;
	if (_Vangle > 360) _Vangle -= 360;
	if (_Vangle < 0) _Vangle += 360;
	//zoom
	if(_isPlusKeyPressed == true && _z < -5.0f) _z = _z + 0.5f;
	if(_isMinKeyPressed == true && _z > -50.0f) _z = _z - 0.5f;
	//move vertical
	if(_isSKeyPressed == true && _y < 8.0f) _y = _y + 0.1f;
	if(_isWKeyPressed == true && _y > -8.0f) _y = _y - 0.1f;
	//move horizontal
	if(_isAKeyPressed == true && _x < 8.0f) _x = _x + 0.1f;
	if(_isDKeyPressed == true && _x > -8.0f) _x = _x - 0.1f;
	//if your turn:
	if(_isAanzet == true) {
	 if(_isSpaceKeyPressed == true) {
	  _isSpaceKeyPressed = false;
	  _pawns[_Zet][_Position] = _pawns[_Zet][_Position] + 1;
	  if(_pawns[_Zet][_Position] == 3) _pawns[_Zet][_Position] = 0;
	 }
	 if(_isEnterKeyPressed == true) {
	  _isEnterKeyPressed = false;
	  if(_Position == 3) {
	   _Zet = _Zet + 1;
	   _isAanzet = false;
	   _Position = 0;
	   if(_pawns[_Zet - 1][0] == 2 && _pawns[_Zet - 1][1] == 2 && _pawns[_Zet - 1][2] == 2 && _pawns[_Zet - 1][3] == 2) {
		//computer heeft gewonnen
		//exit(0);
		_isAanzet == false;
	   } else {
		if(_Zet >= _Rows) {
	     //computer heeft verloren
	    	 _isAanzet == false;
	    } else {
		 addAnswer();
	    }
	   }
	  } else {
	   _Position = _Position + 1;
	  }
	 }
	 if(_isBackspaceKeyPressed == true) {
	  _isBackspaceKeyPressed = false;
	  if(_Position > 0) _Position = _Position - 1;
	 }
	}
	glutPostRedisplay();
	glutTimerFunc(25, update, 0);
}

int main(int argc, char** argv) {

	addAnswer(); //to start

	glutInit(&argc, argv);
	glutInitDisplayMode(GLUT_DOUBLE | GLUT_RGB | GLUT_DEPTH);
	glutInitWindowSize(600, 600);
	
	glutCreateWindow("Mastermind");
	initRendering();
	
	glutDisplayFunc(drawScene);
	glutKeyboardFunc(handleKeypress);
	glutKeyboardUpFunc(handleKeyReleased);
	glutSpecialFunc(handleSpecialKeypress);
	glutSpecialUpFunc(handleSpecialKeyReleased);
	glutReshapeFunc(handleResize);

	glutTimerFunc(25, update, 0);
	glutMainLoop();
	return 0;
}
